src/libcore/ptr.rs

```diff
@@ -1070,13 +1079,40 @@ impl<T: ?Sized> Shared<T> {
     pub unsafe fn new(ptr: *const T) -> Self {
         Shared { pointer: NonZero::new(ptr), _marker: PhantomData }
     }
-}
 
-#[unstable(feature = "shared", issue = "27730")]
-impl<T: ?Sized> Shared<T> {
+    /// Acquires the underlying `*const` pointer.
+    pub fn ptr(self) -> *const T {
+        self.pointer.get() as *const _
+    }
+
+    /// Acquires the underlying `*const` pointer as a `*mut`.
+    pub fn ptr_mut(self) -> *mut T {
+        self.pointer.get() as *mut _
+    }
+
+    /// Dereferences the content.
+    ///
+    /// The resulting lifetime is bound to self so this behaves "as if"
+    /// it were actually an instance of T that is getting borrowed. If a longer
+    /// (unbound) lifetime is needed, use `&*my_ptr.ptr()`.
+    pub unsafe fn as_ref(&self) -> &T {
+        &*self.ptr()
+    }
+
+    /// Mutably dereferences the content.
```

Comment:
Should (IMO) have a warning that this should not be used in conditions of "true sharing" but rather when the access is *known to be unique* through various means (just created `Shared`, a ref-count check, etc.) and that interior mutability (`UnsafeCell` or better) is the more general approach.